core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -424,79 +466,47 @@ protected boolean checkBlocks() {
                 }
                 return false;
             }
+            // request does not have a shardId yet, we need to pass the concrete index to resolve shardId
+            resolveRequest(clusterState.metaData(), concreteIndex, request);
+            assert request.shardId() != null : "request shardId must be set in resolveRequest";
             return true;
         }
 
-        protected ShardRouting resolvePrimary(ShardIterator shardIt) {
-            // no shardIt, might be in the case between index gateway recovery and shardIt initialization
-            ShardRouting shard;
-            while ((shard = shardIt.nextOrNull()) != null) {
-                // we only deal with primary shardIt here...
-                if (shard.primary()) {
-                    return shard;
-                }
-            }
-            return null;
-        }
+        private void performAction(final String nodeId, final String action, final boolean isPrimaryAction) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            transportService.sendRequest(node, action, request, transportOptions, new BaseTransportResponseHandler<Response>() {
```

Comment:
Okay, you're correct. Can we make this clearer? Grab a copy of the observed cluster state at the start of doRun, pass it along to the global checks and request resolving, and finally on its merry way to the rest of the request handling? On a retry, that whole chain will start over after a new observed state is captured?